File: value-harness.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unistd.h>
#include <arpa/inet.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <fstream>
#include <filesystem>
#include <ctime>
#include <cstring>
#include "tuning.pb.h"

int main(int argc, char* argv[]) {
    std::vector<char> inputBuf((std::istreambuf_iterator<char>(std::cin)),
                                std::istreambuf_iterator<char>());

    std::vector<double> values(7);
    std::memcpy(values.data(), inputBuf.data(), 56);

    std::vector<std::string> keys = {
        "speed",
        "kp",
        "kd",
        "ki",
        "servo-trim",
        "threshhold-value",
        "servo-scaler"
    };

    protobuf_msgs::TuningState tuning;
    for (size_t i = 0; i < 7; ++i) {
        auto* param = tuning.add_dynamicparameters();
        auto* numberParam = new protobuf_msgs::TuningState::Parameter::NumberParameter();
        numberParam->set_key(keys[i]);
        numberParam->set_value(values[i]);
        param->set_allocated_number(numberParam);
    }

    std::string serialized;
    if (!tuning.SerializeToString(&serialized)) {
        return 1;
    }

    std::string stateFile = "pipeline_status.flag";
    bool wasConnectedBefore = false;
    if (std::ifstream in{stateFile}) {
        char c;
        in >> c;
        wasConnectedBefore = (c == '1');
    }
 

    int sock = socket(AF_INET, SOCK_STREAM, 0);
    if (sock < 0) {
        return 1;
    }

    sockaddr_in serv_addr{};
    serv_addr.sin_family = AF_INET;
    serv_addr.sin_port = htons(9000);
    if (inet_pton(AF_INET, "192.168.0.146", &serv_addr.sin_addr) <= 0) {
        close(sock);
        return 1;
    }

    if (connect(sock, (sockaddr*)&serv_addr, sizeof(serv_addr)) < 0) {
        std::ofstream(stateFile) << "0";
        if (wasConnectedBefore) {
            std::filesystem::create_directory("crashes");

            std::string crashFile = "crashes/crash_" + std::to_string(std::time(nullptr)) + ".bin";
            std::string scpCommand = "sshpass -p 'debix' scp debix@192.168.0.146:/tmp/last_input.bin " + crashFile;
            std::system(scpCommand.c_str());
        } else {
            system("./start_pipeline.sh");
        }return 1;
    }

    std::ofstream(stateFile) << "1";

    ssize_t total_sent = 0;
    ssize_t to_send = serialized.size();
    const char* buffer = serialized.data();
    while (total_sent < to_send) {
        ssize_t sent = send(sock, buffer + total_sent, to_send - total_sent, 0);
        if (sent <= 0) {
            close(sock);
            return 1;
        }
        total_sent += sent;
    }

    close(sock);

    // std::ofstream out("last_input.bin", std::ios::binary);
    // out.write(inputBuf.data(), inputBuf.size());
    // out.close();

    return 0;
}

File: value-harness.go
// package main

// import (
// 	"log"
// 	"net"
// 	"os"
// 	"strconv"

// 	"syscall"
// 	"time"

// 	//"vu/ase/transceiver/src/serverconnection"
// 	//"vu/ase/transceiver/src/state"
// 	pb_tuning "github.com/VU-ASE/rovercom/packages/go/tuning"
// 	"google.golang.org/protobuf/proto"
// )

// func main() {
// 	// fuzzArg := os.Args[1]

// 	// value, err := strconv.ParseFloat(fuzzArg, 32)
// 	// if err != nil {
// 	// 	log.Fatalf("Error converting argument to float: %v", err)
// 	// }

// 	// // Convert float64 to float32
// 	// fuzzValue := float32(value)

// 	fuzzArg := ""
// 	if len(os.Args) > 1 {
// 		fuzzArg = os.Args[1]
// 	} else {
// 		// If no argument is provided, read from stdin
// 		data := make([]byte, 256) // Adjust size as needed
// 		n, err := os.Stdin.Read(data)
// 		if err != nil {
// 			log.Fatalf("Error reading from stdin: %v", err)
// 		}
// 		fuzzArg = string(data[:n-1])
// 	}

// 	value, err := strconv.ParseFloat(fuzzArg, 32)
// 	if err != nil {
// 		log.Fatalf("Error converting argument to float: %v", err)
// 	}

// 	// Convert float64 to float32
// 	fuzzValue := float32(value)

// 	tuning := &pb_tuning.TuningState{
// 		Timestamp: uint64(time.Now().UnixMilli()),
// 		DynamicParameters: []*pb_tuning.TuningState_Parameter{
// 			{
// 				Parameter: &pb_tuning.TuningState_Parameter_Number{
// 					Number: &pb_tuning.TuningState_Parameter_NumberParameter{
// 						Key:   "speed",
// 						Value: fuzzValue,
// 					},
// 				},
// 			},
// 		},
// 	}
// 	syscall.Kill(syscall.Getpid(), syscall.SIGSEGV)

// 	data, err := proto.Marshal(tuning)
// 	if err != nil {
// 		log.Fatalf("Failed to marshal tuning: %v", err)
// 	}

// 	conn, err := net.Dial("tcp", "192.168.0.146:9000")
// 	if err != nil {
// 		log.Fatalf("Failed to connect to transceiver: %v", err)
// 	}
// 	defer conn.Close()

// 	_, err = conn.Write(data)
// 	if err != nil {
// 		log.Fatalf("Failed to send data: %v", err)
// 	}

// 	log.Printf("TuningState sent to transceiver successfully.")
// }


package main

import (
    "encoding/binary"
    "fmt"
    "io"
    "io/ioutil"
    "net"
    "os"
    "path/filepath"
    "strconv"
    "time"
	"math"

    // Import your generated protobuf package here
    pb "github.com/VU-ASE/rovercom/packages/go/tuning"
    "google.golang.org/protobuf/proto"
)

func main() {
    // Read 56 bytes from stdin
    inputBuf := make([]byte, 28)
    n, err := io.ReadFull(os.Stdin, inputBuf)
    if err != nil {
        fmt.Fprintf(os.Stderr, "Failed to read input: %v\n", err)
        os.Exit(1)
    }
    if n != 28 {
        fmt.Fprintf(os.Stderr, "Expected 56 bytes (7 doubles), got %d\n", n)
        os.Exit(1)
    }

    // Parse 7 float64 values (little endian)
    values := make([]float32, 7)
    for i := 0; i < 7; i++ {
        values[i] = math.Float32frombits(binary.LittleEndian.Uint32(inputBuf[i*4 : (i+1)*4]))
	}
    

    keys := []string{
        "speed",
        "kp",
        "kd",
        "ki",
        "servo-trim",
        "threshhold-value",
        "servo-scaler",
    }

    // Construct TuningState protobuf
	dynamicParams := make([]*pb.TuningState_Parameter, 7)
	for i := 0; i < 7; i++ {
	dynamicParams[i] = &pb.TuningState_Parameter{
			Parameter: &pb.TuningState_Parameter_Number{
				Number: &pb.TuningState_Parameter_NumberParameter{
					Key:   keys[i],
					Value: values[i],
				},
			},
		}
	}		
    tuning := &pb.TuningState{
		DynamicParameters: dynamicParams,
	}
    // for i := 0; i < 7; i++ {
    //     param := &pb.TuningState_Parameter{}
    //     numberParam := &pb.TuningState_Parameter_NumberParameter{
    //         Key:   keys[i],
    //         Value: values[i],
    //     }
    //     param.Param = &pb.TuningState_Parameter_Number{Number: numberParam}
    //     tuning.DynamicParameters = append(tuning.DynamicParameters, param)
    // }

    // Serialize protobuf
    serialized, err := proto.Marshal(tuning)
    if err != nil {
        fmt.Fprintf(os.Stderr, "Failed to serialize protobuf message: %v\n", err)
        os.Exit(1)
    }

    // Handle connection state
    stateFile := "pipeline_status.flag"
    wasConnectedBefore := false
    if data, err := ioutil.ReadFile(stateFile); err == nil && len(data) > 0 {
        wasConnectedBefore = (data[0] == '1')
    }

    // Connect to server
    conn, err := net.Dial("tcp", "192.168.0.146:9000")
    if err != nil {
        _ = ioutil.WriteFile(stateFile, []byte("0"), 0644)
        // _ = startPipeline("./start_pipeline.sh")

        if wasConnectedBefore {
            _ = os.MkdirAll("crashes", 0755)
            crashFile := filepath.Join("crashes", "crash_"+strconv.FormatInt(time.Now().Unix(), 10)+".bin")
            if err := os.Rename("last_input.bin", crashFile); err != nil {
                fmt.Fprintf(os.Stderr, "Failed to save crash input: %v\n", err)
            }
            // Here you would pause/resume AFL++ as needed
            panic(42)
        } else {
            os.Exit(1)
        }
    }
    defer conn.Close()
    _ = ioutil.WriteFile(stateFile, []byte("1"), 0644)

    // Send serialized protobuf message
    totalSent := 0
    toSend := len(serialized)
    for totalSent < toSend {
        n, err := conn.Write(serialized[totalSent:])
        if err != nil {
            fmt.Fprintf(os.Stderr, "send: %v\n", err)
            os.Exit(1)
        }
        totalSent += n
    }

    // Save input to last_input.bin
    if err := ioutil.WriteFile("last_input.bin", inputBuf, 0644); err != nil {
        fmt.Fprintf(os.Stderr, "Failed to write last_input.bin: %v\n", err)
        os.Exit(1)
    }
}

// startPipeline runs a shell script.
// func startPipeline(path string) error {
//     return execShell(path)
// }

// func execShell(path string) error {
//     // Simple implementation; for more control use os/exec
//     return os.system("/bin/sh", "-c", path)
// }
